### ISSUE TITLE
Feature: One click download

### ISSUE THREAD

**Issue by TobiasKe** [Langenscheiss/bibitnow#32]
Hi, I recently stumbled upon this extension. It's great for retrieving citations. At work I have to use a version of Endnote, where it's most convenient to download and open a citation file (e.g. in .ris format).

While it's already great as it is, I would suggest to add an optional feature similar to "auto copy to clipboard" that automatically starts the download of the default format. It'd save me an additional click everytime. Yes, I'm lazy, but I really think it would improve my user experience even more.

Keep up the great work.
Tobias

**Comment by Langenscheiss** [comment]
Hi.
I will look into it. Similarly to the autocopy option, it will probably ask additional permissions from the user, but since the permission system allows to do that gracefully without forcing anything onto the user. If it is possible, I will do it just like for the autocopy, with an additional shortcut.

Question: Would you like to get the download dialogue where you can choose whether to open/download, or do you want the download to start right away?

**Comment by TobiasKe** [comment]
Personally, I ask all my browsers to always ask me, what to do with it.
In this case, I then use the option to open it, as I don't need it saved anywhere but want to open it directly with EndNote to insert it into a library.
But I imagine this might be something every user should be able to decide for themselves. So I guess, you could probably use the default settings of the browser here.

PS: It's awesome how commited you are to adding the new features people request.

**Comment by Langenscheiss** [comment]
Ok. I also think this should be decided by the user.
I guess the way to do it with browser default settings is to have the download button click event triggered automatically I suppose. But it could be that the download API also allows to override this behavior altogether and have it download or open depending on how the user sets this up in the extension options. At least in Firefox, need to check about Chrome. My aim is generally to offer as many customization as reasonably possible while keeping a reasonable default that works for most of the people.

**Comment by Langenscheiss** [comment]
Hi.
If you are Chrome user, I am glad to announce that already version 0.897 will allow you to do what you want. If you are firefox user, I have some semi-bad or, depending on your point of view, only semi-good news. Namely, for the time being, you still need to click once (i.e., one less than before, which is already better according to your lazyness scale, but not perfect and not what I want)

First the technical reason:
So the auto-download via shortcut can in principle be implemented with the downloads API. The problem comes when you actually want to directly open the file with the associated application. Due to obvious security concerns, this is programmatically only possible in a direct user-input handler, i. e., a function that is called directly by user input (such as pressing the keyboard short cut). Unfortunately, the extension works a bit different due to the structure of webextensions. Namely, once you initiate the plugin, it first injects a content script into the website it needs to read, then instructs this script to read, and then waits for the response from the content script. This process is handled asynchronously, meaning that the handling of the response from the content script is done by a different handler, which (as far as I am aware), is not recognized as a user input handler, despite the fact that it was implicitly triggered by the user. So once the program has reached this state, the downloads API only allows you to download, but not to actually open the downloaded content.

In principle, and luckily, HTML5 brings a simple and elegant solution to this: the "download" attribute of html anchor elements (links). If this property is set, you can programmatically trigger the link action without further user input (such as clicking the link with the mouse), because it is assumed that the instructed "download" process is handled securely by the browser. However, Mozilla Firefox and Google Chrome have a different "opinion" about what is secure. Namely, once you tell Google Chrome to process such links by opening the downloaded file with the default application, it will do so after the download. This is what you want. However, as far as I can tell, Firefox ALWAYS asks what to do if you trigger a link with a "download" property, even if you explicitly tell it not to, and even if you actually click the link with the mouse (in the latter case, security is a somewhat lame excuse, as it is as secure as the usual way of dealing with links without download property). You can test this here:

https://www.webfx.com/blog/images/assets/cdn.sixrevisions.com/0435-01_html5_download_attribute_demo/html5download-demo.html

The demo works as intended on Google Chrome, meaning you can set the browser to always open the links with whatever default program you have chosen in the system. On Firefox, it will always ask what to do. I don't know whether this is a bug, but it's the unfortunate truth. Unless there is some neat loophole in this logic using iframes or something similar (still looking into this), completely eliminating this issue probably requires a deep restructuring of the core logic of the extension. At the moment I am simply not willing to do this as it is too time consuming, and I have not yet made certain that it will even bring the desired effect.

So TL:DR: on Firefox, for some time from now, you will probably (if I don't find any other solution) have 2 ways to download with one click less:
1. Open popup as usual, click the link. Now, the link will not have the download property set, so it will actually remember the default download action, and it will not ask again what to do if you tell it not to. This is in particular the best way if you realize that you first need to change citation format or its options.
2. Open the popup using the new shortcut. Now, you will, as soon as the citation is ready, get a download dialogue on which you need to click OK to perform the action. 
In both cases, it is one short-cut and one click. The last click is, unfortunately, not easily avoidable :(

**Comment by Langenscheiss** [comment]
Closed until further notice, since no solution for Firefox is in sight (for me at least)
